django/contrib/auth/models.py

```diff
@@ -368,7 +388,20 @@ class Meta:
 
     def clean(self):
         super().clean()
-        self.email = self.__class__.objects.normalize_email(self.email)
+        # Try the normalize_email implementation on UserManager:
+        # if it raises RemovedInDjango41Warning, we can call the
+        # new implementation here instead. Otherwise, it was
+        # overridden, and we must warn about the deprecation.
```

Comment:
Don't do this. Instead use an `is` check to see if the method has been overridden.

It's a bad idea to use `catch_warnings` in production code since it modifies the warning filters in a non-threadsafe fashion. Muted warnings could be raised whilst this block runs.

```python
manager_normalize_email = self.objects.normalize_email
if manager_normalize_email is not BaseUserManager.normalize_email:
    self.email = manager_normalize_email(self.email)
else:
    self.email = self.normalize_email(self.email)
```

Also the comments on the ticket suggested putting this check inside the new class method, rather than here.

Reply:
This is stick-simple, I love it. (This is neither here nor there, but I ended up going down the wrong path by adapting the approach linked in a comment early in the thread.) Thanks!

Reply:
Right - you want to compare the `__func__` attributes of the bound methods, which point to their underlying functions:

```
In [4]: UserManager.normalize_email.__func__ is BaseUserManager.normalize_email.__func__
Out[4]: True
```

Reply:
Amazing, thanks for walking me through that.